### PR TITLE
[Performance] Pool out-of-process protobuf response decoding

### LINKED ISSUE #3487: [Performance] Eliminate full decoded byte[] allocations for out-of-process responses
**Is your feature request related to a problem? Please describe.**

Every out-of-process orchestrator turn decodes the worker's base64 protobuf response with `Convert.FromBase64String()`, producing a new byte array, and then parses that array into the protobuf object graph:

https://github.com/Azure/azure-functions-durable-extension/blob/4a52e5a802660bdb8fed0141d75b9d1e89463d2f/src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs#L142-L160

The entity-batch response path repeats the same pattern:

https://github.com/Azure/azure-functions-durable-extension/blob/4a52e5a802660bdb8fed0141d75b9d1e89463d2f/src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs#L388-L403

The request direction already rents a buffer specifically to avoid putting a `byte[]` allocation for every protobuf request onto the GC:

https://github.com/Azure/azure-functions-durable-extension/blob/4a52e5a802660bdb8fed0141d75b9d1e89463d2f/src/WebJobs.Extensions.DurableTask/ProtobufUtils.cs#L423-L441

### Performance impact

The decoded array is approximately 75% of the base64 string length and is allocated on every isolated/out-of-process orchestration execution and entity batch. Responses over roughly 85 KB produce large-object-heap allocations. Action-heavy orchestration turns, large custom status values, and large entity result batches can therefore add a sizable transient buffer on top of the base64 string and the protobuf object graph that must already exist.

A decode-only .NET 10.0.10 microbenchmark using a 1 MiB decoded payload (40 iterations) measured:

| Decode path | Mean per operation | Allocated per operation |
|---|---:|---:|
| `Convert.FromBase64String` | 2.59 ms | 1,048,601 B |
| Pooled buffer + `Convert.TryFromBase64String` | 1.08 ms | ~1 B |

This benchmark intentionally isolates base64 decoding; it does not include protobuf parsing or worker/host IPC. The expected end-to-end benefit is lower Gen 2/LOH pressure and lower tail latency under concurrent, payload-heavy turns rather than eliminating protobuf graph allocations.

**Are there any existing GitHub discussions or issues filed that help give some context to this proposal?**

I did not find an exact open issue for response-side base64 decode allocation. Large-payload and memory-pressure reports provide broader context, but this proposal does not change payload storage, serialization format, or the language-worker protocol.

**Describe the solution you'd like**

Introduce one internal helper used by both response paths that:

1. Computes the maximum decoded length with checked arithmetic.
2. Rents an appropriately sized buffer from `ArrayPool<byte>.Shared` (with a documented policy for unusually large buffers).
3. Uses `Convert.TryFromBase64String` to decode into that buffer.
4. Parses only the valid `[offset, bytesWritten]` segment using a supported protobuf `CodedInputStream`/parser overload, without first copying it into a right-sized array.
5. Returns the buffer in a `finally` block after parsing completes.
6. Preserves the current malformed-base64 and malformed-protobuf failure behavior; failed decode must never be interpreted as an empty/default response.

Apply the helper to both `OrchestratorResponse` and `EntityBatchResult` parsing. Keep the existing request-side encoding path unchanged unless a shared abstraction clearly reduces duplication.

### Backward-compatibility requirements

- Keep the response contract exactly base64-encoded protobuf; no worker SDK, schema, or wire-protocol change.
- Preserve all parsed actions, custom status, failure details, `RequiresHistory`, and `RequiresState` behavior.
- Preserve the exception/error path for null, empty, malformed base64, and invalid protobuf responses.
- Do not retain the rented buffer or expose memory backed by it after the parser returns.
- Follow an explicit buffer-clearing policy consistent with the existing request-side pool usage and repository security expectations.

**Describe alternatives you've considered**

- **Send raw binary through the Functions worker protocol.** The existing code documents why base64 is used across language workers, and changing it would be a cross-SDK protocol change with significant compatibility risk.
- **Use `MemoryStream` around the base64 string.** This does not avoid the decoded byte storage and generally adds another object/stream layer.
- **Only optimize responses above the LOH threshold.** This reduces implementation risk but leaves high-frequency smaller allocations in place; a shared pooled helper should handle both efficiently.
- **Cache decoded responses.** Responses are turn-specific and caching would increase retention with little reuse.

**Additional context**

### Suggested validation and acceptance criteria

- Add round-trip tests for orchestrator and entity responses at small, pooled, and multi-megabyte sizes.
- Cover padded/unpadded-valid cases as supported today, malformed base64, invalid protobuf, and pool return on every exception path.
- Compare parsed messages field-for-field with the current implementation, including unknown protobuf fields if applicable.
- Add allocation benchmarks at 1 KB, 64 KB, 128 KB, and 1 MiB.
- Acceptance target: eliminate the response-sized managed `byte[]` allocation on the normal path and materially reduce decode time without changing the base64/protobuf contract or observable errors.

### PR DESCRIPTION
**Pull request by berndverst**
# Summary
## What changed?
- Added a shared pooled base64 decode/protobuf parse helper for out-of-process responses.
- Routed both orchestrator and entity batch responses through the helper.
- Added small, 1 MiB, malformed-base64, malformed-protobuf, whitespace, valid-segment, null-argument, and pool-lifetime coverage for both message types.

## Why is this change needed?
- `Convert.FromBase64String` allocated a response-sized managed `byte[]` on every turn, including LOH-sized arrays for large responses.
- The new normal path computes a checked decoded bound, rents from `ArrayPool<byte>.Shared`, decodes with `Convert.TryFromBase64String`, and parses only `[0, bytesWritten]`, eliminating that full decoded-array allocation while retaining the protobuf object graph required by the response.

## Issues / work items
- Fixes #3487

## Compatibility guarantees
- The contract remains the same base64-encoded protobuf bytes; there is no worker SDK, schema, public API, or wire-format change.
- Existing action, custom-status, failure-details, `RequiresHistory`, and `RequiresState` handling is unchanged.
- Existing orchestrator/entity null and empty response guards are unchanged.
- Malformed base64 falls back to `Convert.FromBase64String` after returning the pool buffer, preserving its exception behavior; malformed protobuf uses the same parser and exception behavior.
- Parsing completes before the buffer is returned, only the valid decoded segment is exposed to protobuf, and no pooled memory is retained.
- Buffer clearing matches the existing request-side `Base64Encode` policy (`clearArray: false`).

---

# Project checklist
- [x] Documentation changes are not required
- [x] Release notes are not required for the next release
- [x] Backport is not required
- [x] All required tests have been added/updated (unit tests, E2E tests)
- [x] No extra work is required to be leveraged by OutOfProc SDKs
- [x] No change to the version of the `WebJobs.Extensions.DurableTask` package
- [x] No EventIds were added to `EventSource` logs
- [ ] This change should be added to the `v2.x` branch
  - [x] This change applies exclusively to `WebJobs.Extensions.DurableTask` v3.x and will be retained only in the `dev` and `main` branches
- [x] Breaking change? No

---

# AI-assisted code disclosure (required)
## Was an AI tool used? (select one)
- [ ] No
- [ ] Yes, AI helped write parts of this PR (e.g., GitHub Copilot)
- [x] Yes, an AI agent generated most of this PR

If AI was used:
- Tool(s): GitHub Copilot CLI
- AI-assisted areas/files: pooled decode helper, middleware wiring, and focused unit tests
- What you changed after AI output: verified repository history and Google.Protobuf 3.21.5 APIs, reviewed whitespace/error/lifetime edge cases, addressed review findings, and validated both target frameworks

AI verification (required if AI was used):
- [x] I understand the code and can explain it
- [x] I verified referenced APIs/types exist and are correct
- [x] I reviewed edge cases/failure paths (timeouts, retries, cancellation, exceptions)
- [x] I reviewed concurrency/async behavior
- [x] I checked for unintended breaking or behavior changes

---

# Testing
## Automated tests
- `dotnet test test\FunctionsV2\WebJobs.Extensions.DurableTask.Tests.V2.csproj -f net8.0 --no-restore --filter "FullyQualifiedName~ProtobufUtilsTests|FullyQualifiedName~OutOfProcMiddlewareTests" --logger "console;verbosity=minimal"` — Passed: 57, Failed: 0, Skipped: 0.
- `dotnet test test\FunctionsV2\WebJobs.Extensions.DurableTask.Tests.V2.csproj -f net10.0 --no-restore --filter "FullyQualifiedName~ProtobufUtilsTests|FullyQualifiedName~OutOfProcMiddlewareTests" --logger "console;verbosity=minimal"` — Passed: 57, Failed: 0, Skipped: 0.
- `dotnet build src\WebJobs.Extensions.DurableTask\WebJobs.Extensions.DurableTask.csproj -c Release --no-restore --verbosity:minimal` — Succeeded for `net8.0` and `net10.0`; 0 warnings, 0 errors.
- `dotnet build test\FunctionsV2\WebJobs.Extensions.DurableTask.Tests.V2.csproj -c Release --no-restore --verbosity:minimal` — Succeeded for `net8.0` and `net10.0`; 0 warnings, 0 errors.

## Manual validation (only if runtime/behavior changed)
- N/A; the wire contract and externally observable behavior are unchanged.

---

# Notes for reviewers
- Dependency inspection confirmed `Google.Protobuf` 3.21.5 exposes `MessageParser<T>.ParseFrom(byte[], int, int)` on both target frameworks.

### CODE REVIEW

File: test/FunctionsV2/Tests/Unit/ProtobufUtilsTests.cs [hunk 1]
```diff
@@ -64,6 +67,148 @@ public void ToOperationAction_StartNewOrchestration_AppliesDefaultVersion(
...
+            if (orchestratorResponse)
+            {
+                actualException = Assert.Throws<FormatException>(
+                    () => ProtobufUtils.Base64Decode("not-base64", P.OrchestratorResponse.Parser, pool));
+            }
+            else
+            {
+                actualException = Assert.Throws<FormatException>(
+                    () => ProtobufUtils.Base64Decode("not-base64", P.EntityBatchResult.Parser, pool));
+            }
```
[the remark this answers is not in the document]
Comment (berndverst) [reply test/FunctionsV2/Tests/Unit/ProtobufUtilsTests.cs]: Addressed in 2181471e. The generated orchestrator and entity parsers have different generic types, so I selected a type-safe `Action` with the conditional expression and now use a single `Assert.Throws<FormatException>` assignment. The behavior and assertions are unchanged; the focused test passes on both net8.0 and net10.0.

File: test/FunctionsV2/Tests/Unit/ProtobufUtilsTests.cs [hunk 2]
```diff
@@ -507,5 +645,50 @@ public void ToPurgeInstanceFilter_EmptyStatusList_MapsToNull()
...
+            public override void Return(byte[] array, bool clearArray = false)
+            {
+                Assert.Same(this.rentedBuffer, array);
+                this.ReturnCount++;
+                this.ClearArray = clearArray;
+
+                // Overwrite returned storage so parsed messages cannot rely on pooled memory.
+                Array.Fill(array, (byte)0);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (berndverst) [reply test/FunctionsV2/Tests/Unit/ProtobufUtilsTests.cs:712]: Addressed in 3192c13a. `TrackingByteArrayPool.Return` now clears its active rental reference after overwriting the returned array, so the test pool supports subsequent Rent/Return cycles without weakening any existing lifetime assertions.

File: src/WebJobs.Extensions.DurableTask/ProtobufUtils.cs [hunk 3]
```diff
@@ -441,6 +441,53 @@ public static string Base64Encode(IMessage message)
...
+            return Base64Decode(encodedMessage, parser, ArrayPool<byte>.Shared);
+        }
+
+        internal static T Base64Decode<T>(
+            string encodedMessage,
+            MessageParser<T> parser,
+            ArrayPool<byte> bufferPool)
+            where T : IMessage<T>
+        {
+            ReadOnlySpan<char> encodedSpan = encodedMessage.AsSpan();
```
[the remark this answers is not in the document]
Comment (berndverst) [reply src/WebJobs.Extensions.DurableTask/ProtobufUtils.cs:460]: Addressed in 3192c13a. The shared overload now validates `encodedMessage`, `parser`, and `bufferPool` before sizing or renting, with focused assertions for the parameter names and no pool activity on invalid input. The middleware null/empty response behavior remains unchanged.

**Comment by berndverst** [comment]
## Decode/parse performance measurements

I measured the exact pre-PR response path against this PR's helper:

```csharp
// Baseline
Parser.ParseFrom(Convert.FromBase64String(encoded));

// PR
ProtobufUtils.Base64Decode(encoded, Parser);
```

**Code points:** baseline/merge-base [`4a52e5a802660bdb8fed0141d75b9d1e89463d2f`](https://github.com/Azure/azure-functions-durable-extension/commit/4a52e5a802660bdb8fed0141d75b9d1e89463d2f), PR head [`3192c13af34f1d6081794c887491367ef1ad0042`](https://github.com/Azure/azure-functions-durable-extension/commit/3192c13af34f1d6081794c887491367ef1ad0042). Current `dev` was `08f8877d0a0f1a8a221a9f6a870e6fb78fb89f92`; I did not rebase or resolve the current conflict, and made no production changes for this measurement.

### Scenarios

- **Worst steady-state CPU:** a nearly empty response (`OrchestratorResponse` 3 B; `EntityBatchResult` 6 B), where pool lookup/return can cost more than allocating the tiny decoded array.
- **Average/representative (not fleet-weighted):** an orchestration turn with four scheduled-task actions (2,229 B), and an entity batch with eight results plus signal/start actions (3,288 B).
- **Best realistic:** a 128 KiB custom-status/entity-state response (over the ~85 KB LOH threshold), and a response just over 1 MiB.
- **Whitespace compatibility stress:** the typical payload with valid CRLF every 76 base64 characters. Workers normally return compact base64; this isolates the accepted-whitespace path.

### Results

Times are **median [min–max]** across 15 measured iterations after 6 warmups. Ops/s is derived from the median. `Allocated/op` is the warm-pool managed allocation reported by MemoryDiagnoser. The final column is the separately measured allocation for the first rent of that target `ArrayPool.Shared` bucket in a fresh process.

| Message / scenario | Decoded | Baseline → pooled time | Baseline → pooled ops/s | Allocated/op (saved) | Fresh first-bucket rent |
|---|---:|---:|---:|---:|---:|
| Orchestrator / tiny | 3 B | 93.8 [91.3–96.3] → 107.4 [105.9–109.8] ns | 10.66M → 9.32M (**0.87x**) | 344 → 312 B (-32 B) | +40 B (16 B bucket) |
| Entity / tiny | 6 B | 121.0 [119.9–123.2] → 128.4 [124.7–131.9] ns | 8.26M → 7.79M (**0.94x**) | 408 → 376 B (-32 B) | +40 B (16 B bucket) |
| Orchestrator / typical | 2,229 B | 5.325 [5.208–5.617] → 3.814 [3.778–3.839] us | 187.8K → 262.2K (**1.40x**) | 8,224 → 5,968 B (-2,256 B) | +4,120 B (4 KiB bucket) |
| Entity / typical | 3,288 B | 8.167 [7.889–8.414] → 5.753 [5.677–5.824] us | 122.4K → 173.8K (**1.42x**) | 11,448 → 8,136 B (-3,312 B) | +4,120 B (4 KiB bucket) |
| Orchestrator / typical + CRLF | 2,229 B | 6.130 [6.065–6.175] → 6.052 [6.003–6.083] us | 163.1K → 165.2K (**1.01x**) | 8,224 → 5,968 B (-2,256 B) | +4,120 B |
| Entity / typical + CRLF | 3,288 B | 9.295 [9.235–9.424] → 9.086 [9.016–9.171] us | 107.6K → 110.1K (**1.02x**) | 11,448 → 8,136 B (-3,312 B) | +4,120 B |
| Orchestrator / 128 KiB | 131,132 B | 375.7 [350.2–382.1] → 236.3 [217.9–240.7] us | 2,662 → 4,233 (**1.59x**) | 394,228 → 263,040 B (-131,188 B) | +262,168 B (256 KiB bucket) |
| Entity / 128 KiB | 131,110 B | 353.6 [329.6–356.1] → 235.8 [226.4–241.3] us | 2,828 → 4,240 (**1.50x**) | 393,996 → 262,832 B (-131,164 B) | +262,168 B |
| Orchestrator / ~1 MiB | 1,048,636 B | 2.539 [2.497–2.589] → 1.442 [1.428–1.449] ms | 394 → 693 (**1.76x**) | 3,146,798 → 2,098,174 B (-1,048,624 B) | +2,097,176 B (2 MiB bucket) |
| Entity / ~1 MiB | 1,048,614 B | 2.335 [2.300–2.386] → 1.433 [1.425–1.454] ms | 428 → 698 (**1.63x**) | 3,146,629 → 2,098,342 B (-1,048,287 B) | +2,097,176 B |

### Interpretation

- **Average/representative:** the warm pooled path used **28–30% less decode/parse CPU**, delivered **1.40–1.42x throughput**, and removed **2.2–3.3 KiB of Gen0 allocation per turn**. At an illustrative 1,000 turns/s, that is 2.3–3.3 MB/s less managed allocation; this is arithmetic, not an assumed production traffic rate.
- **Best measured:** 128 KiB–1 MiB responses were **1.50–1.76x faster** in this decode+parse microbenchmark and removed one response-sized allocation per turn. At 100 1-MiB turns/s, the allocation reduction alone is about 105 MB/s. The 128 KiB cases measured Gen2/full collections falling from 124.5 to 83.0–83.3 per 1,000 operations (~33%); importantly, large parsed strings/object graphs still allocate on the LOH, so this does **not** eliminate LOH/Gen2 activity.
- **Worst steady-state CPU:** tiny responses regressed **6.1% (entity) to 14.5% (orchestrator)**, but the absolute cost was only **7–14 ns/turn**, while still saving 32 B. Valid whitespace largely erased the CPU gain (only 1–2%) but retained the allocation reduction.
- **Worst cold/concurrent memory:** pooling trades repeated allocation/GC for bucket-rounded reusable storage. Across 21 fresh processes per case, the first target-bucket allocation delta was invariant: 40 B, 4,120 B, 262,168 B, or 2,097,176 B. Since the ~1 MiB wire payload is just over the 1 MiB bucket boundary, a cold 32-way burst can allocate/retain roughly **64 MiB of 2 MiB pool buffers**, versus roughly **32 MiB of transient decoded arrays** on the old path (the common protobuf graph is additional in both). `ArrayPool.Shared` can later reuse or trim these buffers. Fresh-process first-call wall times ranged 0.294–4.374 ms, but they include cold caches/tiering despite cross-bucket pre-JIT, so I do not attribute that timing solely to the pool.
- The 1 MiB MemoryDiagnoser run reported normalized Gen2 counts increasing from 218.75 to 281.25/1,000 even with ~1 MiB less allocation/op, likely reflecting retained 2 MiB pool buffers and benchmark batching. That is why I treat object/allocation removal as the reliable GC evidence and do not project a fleet Gen2-rate reduction from this microbenchmark.

### Methodology and caveats

- Scratch-only signed friend project (not committed), referencing the PR-head product project and using the actual generated `OrchestratorResponse` / `EntityBatchResult` types and `Google.Protobuf` parser.
- The only files changed between the measured merge-base and head are `OutOfProcMiddleware.cs`, `ProtobufUtils.cs`, and `ProtobufUtilsTests.cs`; the protobuf schema/package is identical. Running both expressions in the same process therefore isolates the response decode-path change.
- BenchmarkDotNet 0.15.8, Release, .NET 10.0.10 / SDK 10.0.302, x64 RyuJIT, concurrent workstation GC, Windows 11 Hyper-V, AMD EPYC 7763 (16 physical / 32 logical cores). One launch, 6 warmups, 15 measured iterations, MemoryDiagnoser, min/max/median columns.
- Every setup and every cold-pool trial verified that both parsed messages were field-for-field equal to the constructed message. Cold-pool allocation used 21 independent fresh processes per message/scenario and the production `ArrayPool.Shared` path.
- Reproduction shape: `[MemoryDiagnoser]` + `[SimpleJob(launchCount: 1, warmupCount: 6, iterationCount: 15)]`, with the two benchmark methods shown above; payloads are described under **Scenarios**. The compact-base64 and CRLF cases use identical decoded protobuf bytes.
- This measures **only base64 decode + protobuf parse**. It excludes worker IPC, the already-existing base64 string, action conversion, orchestration/entity execution, storage, and host scheduling. Therefore the measured CPU delta is a per-turn decode contribution, **not** an end-to-end latency claim. Absolute timings are machine/runtime-specific; relative allocation counts and the one-fewer-decoded-array result are the more portable conclusions.
